### PR TITLE
Remove the service worker and ui.USE_SERVICE_WORKER option

### LINKED ISSUE #7372: reference error in PWA serviceworker
`ReferenceError: caches is not defined` - serviceworker.js, line 60.

Appeared in #4730 PR, reproducible on https://try.gitea.io if you load Gitea **in Firefox.** Firefox doesn't have the `caches` variable.

### LINKED ISSUE #8056: serviceworker causes basic auth dialog to not show on subsequent visits
Affects all versions since the serviceworker was introduced presumably, including `1.10.0+dev-232-gf1c414882`.

Problem:
- Run gitea behind any kind of reverse proxy (e.g. nginx) with configured basic auth (in nginx).
- Visit it in browser - the auth dialog will be displayed, enter credentials.
- Close the browser.
- Reopen the browser, visit the page again.
- It shows the "401 Unauthorized" page, without displaying the basic auth dialog - even though the `WWW-Authenticate` header was sent. In Chrome, refreshing the page with `F5` doesn't work. Only pressing `Ctrl+Shift+R` (not possible on mobile obviously) causes the auth dialog to pop up again.

Seems to be related to the PWA / service worker feature. There is plenty of info on this issue on the web, googling `serviceworker basic auth` shows many relevant results.

Older versions of gitea (tested with 1.6, see below) didn't suffer from this problem, due to not having the PWA feature.

The issue can be reproduced by using the docker-compose file at https://gist.github.com/peterholak/317f4ca1827672aa231c7114a19149b5. After running `docker-compose up`, the page at `localhost:8004` (latest version of gitea from docker hub) suffers from the described problem, while the page at `localhost:8005` (version 1.6) does not. (The credentials are user `hello`, password `world`.)

If this is not easy to solve, it would be nice to at least be able to disable the PWA in config.

### LINKED ISSUE #11092: Automate serviceworker caching manifest generation
The script at https://github.com/go-gitea/gitea/blob/master/templates/pwa/serviceworker_js.tmpl should be replaced with something that automatically generates a appropriate service worker caching manifest based on the present files in `public` at compile time and then use that data together with  `StaticUrlPrefix` at runtime to serve the actual URLs.

Normally, one would use a webpack plugin for this, but webpack can not know about `StaticUrlPrefix`, and also does not know about `vendor` files so it's not directly suitable, but could maybe be integrated in the compile-time part of the process.

### PR DESCRIPTION
**Pull request by silverwind**
We don't really gain anything from it except the client-side caching which is a problem better solved on HTTP level. Gitea is not a PWA and will not do anything useful in a offline scenario without having access to the server.

This should also resolve issues seen with the serviceworker interfering on SSE requests.

The only API we can not use without a service worker is push notifications while the "app" is closed but the more established HTML Notification API will be a alternative.

Fixes: https://github.com/go-gitea/gitea/issues/11092
Fixes: https://github.com/go-gitea/gitea/issues/8056
Fixes: https://github.com/go-gitea/gitea/issues/7372

### CODE REVIEW

**Review by zeripath — APPROVED** [review]

**Comment by zeripath** [comment]
Probably this needs more than just two approvals

**Comment by CirnoT** [comment]
Might considering marking this as breaking; I'm sure people that actively use Gitea as PWA (as much as it isn't meant to be) would consider it as such.

**Review by CirnoT — APPROVED** [review]

**Review by mrsdizzie — APPROVED** [review]

**Review by jolheiser — APPROVED** [review]
I wonder if there is any way that this could be added back manually (without code change) for those that want it, through custom content or something similar.  
I may try to play with it and add docs or contrib if this gets merged.

**Review by 6543 — APPROVED** [review]

**Comment by 6543** [comment]
plreace resolve conflict

File: routers/routes/routes.go [hunk 1]
```diff
@@ -1048,10 +1048,6 @@ func RegisterRoutes(m *macaron.Macaron) {
 		ctx.HTML(200, "pwa/manifest_json")
```
Comment (lunny) [inline routers/routes/routes.go:1048]: should we change this also?

**Comment by lunny** [comment]
Why not just change the default option to false?

**Comment by CirnoT** [comment]
> Why not just change the default option to false?

That means that we would still have a broken configuration; specifically with SSE request being killed by SW

**Comment by silverwind** [comment]
I'll close this and follow up with a better approach to the SW (which in itselfs still seems valuable to keep for performance) using [workbox](https://developers.google.com/web/tools/workbox) with [destination-based caching](https://medium.com/dev-channel/service-worker-caching-strategies-based-on-request-types-57411dd7652c#3b5d). This will also free us from the burden of having to manually keep that file up to date.

**Comment by CirnoT** [comment]
>We have it configured as a static list of resources to cache and non-matches should always hit the network.
But the request that hits network is intercepted by SW, and SW are not supposed to be working forever; hence Firefox kills it after 30s, breaking SSE.

**Comment by silverwind** [comment]
As discussed, the SSE issue should be because the SW unconditionally handled all requests. Should be fixed in https://github.com/go-gitea/gitea/pull/11538 which only intercepts a subset of requests.

**Comment by silverwind** [comment]
> Do we really need to additionally burden Gitea with libraries designed for offline usage, such as workbox

I'd say it's justified because when using serviceworkers, we can guarantee at least some degree of caching which is important for performance because we do have some large assets (monaco) that really benefit from it. The current workbox implementation is only 28kB, so very lightweight.

**Pull request closed without merging**